x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushFiltersToSource.java

```diff
@@ -53,47 +59,96 @@
-                    exp
-                );
-            }
-            // Combine GT, GTE, LT and LTE in pushable to Range if possible
-            List<Expression> newPushable = combineEligiblePushableToRange(pushable);
-            if (newPushable.size() > 0) { // update the executable with pushable conditions
-                Query queryDSL = PlannerUtils.TRANSLATOR_HANDLER.asQuery(Predicates.combineAnd(newPushable));
-                QueryBuilder planQuery = queryDSL.asBuilder();
-                var query = Queries.combine(Queries.Clause.FILTER, asList(queryExec.query(), planQuery));
-                queryExec = new EsQueryExec(
-                    queryExec.source(),
-                    queryExec.index(),
-                    queryExec.indexMode(),
-                    queryExec.output(),
-                    query,
-                    queryExec.limit(),
-                    queryExec.sorts(),
-                    queryExec.estimatedRowSize()
-                );
-                if (nonPushable.size() > 0) { // update filter with remaining non-pushable conditions
-                    plan = new FilterExec(filterExec.source(), queryExec, Predicates.combineAnd(nonPushable));
-                } else { // prune Filter entirely
-                    plan = queryExec;
-                }
-            } // else: nothing changes
+            plan = planFilterExec(filterExec, queryExec, ctx);
+        } else if (filterExec.child() instanceof EvalExec evalExec && evalExec.child() instanceof EsQueryExec queryExec) {
+            plan = planFilterExec(filterExec, evalExec, queryExec, ctx);
         }
-
         return plan;
     }
 
+    private static PhysicalPlan planFilterExec(FilterExec filterExec, EsQueryExec queryExec, LocalPhysicalOptimizerContext ctx) {
+        List<Expression> pushable = new ArrayList<>();
+        List<Expression> nonPushable = new ArrayList<>();
+        for (Expression exp : splitAnd(filterExec.condition())) {
+            (canPushToSource(exp, x -> LucenePushDownUtils.hasIdenticalDelegate(x, ctx.searchStats())) ? pushable : nonPushable).add(exp);
+        }
+        return rewrite(filterExec, queryExec, pushable, nonPushable, List.of());
+    }
+
+    private static PhysicalPlan planFilterExec(
+        FilterExec filterExec,
+        EvalExec evalExec,
+        EsQueryExec queryExec,
+        LocalPhysicalOptimizerContext ctx
+    ) {
+        AttributeMap<Attribute> aliasReplacedBy = getAliasReplacedBy(evalExec);
+        List<Expression> pushable = new ArrayList<>();
+        List<Expression> nonPushable = new ArrayList<>();
+        for (Expression exp : splitAnd(filterExec.condition())) {
+            Expression resExp = exp.transformUp(ReferenceAttribute.class, r -> aliasReplacedBy.resolve(r, r));
+            (canPushToSource(resExp, x -> LucenePushDownUtils.hasIdenticalDelegate(x, ctx.searchStats())) ? pushable : nonPushable).add(
```

Comment:
Very subtle bug: in case we don't push the attribute, we mustn't add it to the non pushables!

Since `resExp` comes with resolved reference attributes, it may point to attributes that existed before the eval but are shadowed in the eval. Like so:
```
test has fields x and y:

FROM test | EVAL y = 2*y, dist = st_distance(x, "POINT(0 0)") | WHERE dist < 10 AND y > 1
```
For `y > 1`, resExp uses a different `y` which does not exist anymore after the Eval!